### ISSUE TITLE
[question] How can I customize the configuration of an imported dependency?

### ISSUE THREAD

**Issue by obergner** [conan-io/conan#16004]
### What is your question?

One of my projects uses the  [spdlog/1.13.0](https://conan.io/center/recipes/spdlog?version=1.13.0) library for log output. `spdlog` supports three different ways of formatting log output:

1. Use the external `fmt` library, the default in the conan package, activated by setting `-DSPDLOG_FMT_EXTERNAL`.
2. Use C++ 20 `std::format`, activated by setting the conan package option `use_std_fmt=True`.
3. Use formatting support built into the library, for which no conan package option exists.

My code base works **only** with the third option. Therefore, I'm looking for a way to unset `-DSPDLOG_FMT_EXTERNAL`. How can I do that?

**Detailed Background**

I port a C++ project with a makefile-based build system to CMake (3.29) and Conan (2.2.2). This project depends on the `spdlog` library that I import in my `conanfile.txt`:
```
[requires]
...
spdlog/1.13.0

[generators]
CMakeDeps
CMakeToolchain

[layout]
cmake_layout
```
Following the Conan tutorial, I imported dependencies and generated CMake integration artifacts via
```
conan install . --output-folder=conan --build=missing
```
I then configured my project via
```
cmake -B out -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=conan/build/Release/generators/conan_toolchain.cmake -S .
```
Eventually, I ran
```
cmake --build . -v
```
from the build folder.

However, my build fails because Conan configures `spdlog` to use an external formatting library (`fmt`). To do so, it sets the compiler directive `-DSPDLOG_FMT_EXTERNAL`, in one of the cmake files it generates. To fix my build I need to unset/overwrite this compiler directive.

**What I tried so far**

What I tried so far is
```
target_compile_options(${NAME}
  PRIVATE 
  -DSPDLOG_FMT_EXTERNAL=OFF)
```
However, this simply _adds_ `-DSPDLOG_FMT_EXTERNAL=OFF` _after_ `-DSPDLOG_FMT_EXTERNAL` and thus does not fix my issue.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @obergner 

Thanks for your question.

I think the reason why ``spdlog`` recipe doesn't have an option to use the bundled fmt copy is because this would be using the vendored copy, but the ``fmt`` headers are actually publicly included in the ``spdlog`` public headers, so they are not an implementation detail, they are propagated and visible to the consumers.

This means that if any other package is using directly or indirectly the ``fmt`` it will result in an error, with 2 copies of ``fmt`` in the include path, one the Conan package and the other the one vendored inside ``spdlog``. As a general rule, dependencies cannot be vendored for package managers, because it creates problems, from duplicated headers to duplicated symbols or ODR violations.

Why using the vendored ``fmt`` is the only approach you can use? There shouldn't be any difference, and ``spdlog`` is very widely used by tons of Conan users with the external ``fmt`` transitive dependency without issues.

**Comment by obergner** [comment]
@memsharded Thanks for coming back to me so quickly. Regarding your comment and questions:

1. Using the copy of `fmt` bundled in `spdlog`, and using an external `fmt` are **not** equivalent. I suspect that `spdlog` bundles an old version of `fmt`. Our codebase has been developed using this old copy. It compiles if we use the bundled `fmt` version (in the old makefile-based build where we have full control over every aspect of the build). If we were to use an external, much more recent `fmt` version we would need to fix _lots_ of compilation issues.
2. `spdlog` is the _only_ dependency that uses `fmt`, and due to the nature of our application we don't foresee any other dependency, nor our own code, to use an external `fmt` in the (foreseeable) future. Therefore, in our situation `ODR` violations should not be an issue.

In summary: we could invest several days into fixing compilation issues caused by using an external `fmt`, just to restore what we already have today. Or we could look for a quick way to unset `-DSPDLOG_FMT_EXTERNAL`. We opt for the latter.

**Comment by obergner** [comment]
@memsharded Just to be clear: I'm **not** looking for an option in the `spdlog` conan package that allows using the bundled `fmt` copy. I know that this doesn't exist, and given what you said one may well argue that it _should not_ exist.

What I'm looking for is for a way to unset `-DSPDLOG_FMT_EXTERNAL` defined by one of the cmake files generated by Conan when installing `spdlog`. I'd be a happy camper if I had that.

**Comment by obergner** [comment]
@memsharded Never mind. I managed to remove `-DSPDLOG_FMT_EXTERNAL` by using `conan.tools.files.replace_in_file()` in the `generate()` method. Which buys me exactly nothing since the `spdlog` package has been _installed_ with `-DSPDLOG_FMT_EXTERNAL`. Meaning that the bundled `fmt` copy has been removed :-( I will need to think of something else.

**Comment by memsharded** [comment]
> Using the copy of fmt bundled in spdlog, and using an external fmt are not equivalent. I suspect that spdlog bundles an old version of fmt. Our codebase has been developed using this old copy.

You can ``self.requires("fmt/version_you_want", override=True)`` in your consumer to force ``spdlog`` to use the ``fmt`` version that you want (if it works of course at the code level).

> spdlog is the only dependency that uses fmt, and due to the nature of our application we don't foresee any other dependency, nor our own code, to use an external fmt in the (foreseeable) future. Therefore, in our situation ODR violations should not be an issue.

That is fine, so far it seems that around other 37 packages in ConanCenter depend on ``fmt``, and transitively that would impact even more. You are lucky that you are not using any other dependency that transitively brings ``fmt``.


> In summary: we could invest several days into fixing compilation issues caused by using an external fmt, just to restore what we already have today. Or we could look for a quick way to unset -DSPDLOG_FMT_EXTERNAL. We opt for the latter.

The compilation and build flags for dependencies are private implementation details, it is not possible to override the dependencies builds, beside some use cases, like defining some compile flags via ``tools.build:cxx_flags``, or it is possible to inject extra CMake variables via a cmake toolchain file with ``tools.cmake.cmaketoolchain:user_toolchain``, but these are not capable of overriding options. In this case, if the recipe is defining ``SPDLOG_FMT_EXTERNAL`` via an option, it is not possible to change that value (otherwise it would be inconsistent with the option definition).

I see a couple of possible approaches to this:

- Follow the recommendations in https://docs.conan.io/2/devops/using_conancenter.html, and modify the recipe of ``spdlog`` to customize it for your need, create the package from your fork.
- Try the above suggestion of checking what is the ``fmt`` version that is vendored inside ``spdlog`` and then do an ``override`` to that specific version. That should in theory provide a transparent replacement? In any case, I am not sure that spdlog is bundling such an old version of fmt, I'd say that it is using 9.1, maybe this approach is not that problematic.

**Comment by memsharded** [comment]
> @memsharded Never mind. I managed to remove -DSPDLOG_FMT_EXTERNAL by using conan.tools.files.replace_in_file() in the generate() method. Which buys me exactly nothing since the spdlog package has been installed with -DSPDLOG_FMT_EXTERNAL. Meaning that the bundled fmt copy has been removed :-( I will need to think of something else.

Yes, exactly, that is the issue, the consumer cannot affect that easily the build of the dependencies in that way. I hope my suggestions in the above comment can help.

**Comment by obergner** [comment]
@memsharded Thanks, much appreciated. I will explore both options mentioned by you. Closing this issue now.

**Comment by memsharded** [comment]
Don't hesitate to re-open or create new tickets if you have any further question while checking those approaches. Thanks!
